torch/distributions/mixture_same_family.py

```diff
@@ -0,0 +1,181 @@
+
+
+class MixtureSameFamily(Distribution):
+    r"""
+    The `MixtureSameFamily` distribution implements a (batch of) mixture
+    distribution where all component are from different parameterizations of
+    the same distribution type. It is parameterized by a `Categorical`
+    "selecting distribution" (over `k` component) and a component
+    distribution, i.e., a `Distribution` with a rightmost batch shape
+    (equal to `[k]`) which indexes each (batch of) component.
+
+    Examples::
+
+        # Construct Gaussian Mixture Model in 1D consisting of 5 equally
+        # weighted normal distributions
+        >>> mix = D.Categorical(torch.ones(5,))
+        >>> comp = D.Normal(torch.randn(5,), torch.rand(5,))
+        >>> gmm = MixtureSameFamily(mix, comp)
+
+        # Construct Gaussian Mixture Modle in 2D consisting of 5 equally
+        # weighted bivariate normal distributions
+        >>> mix = D.Categorical(torch.ones(5,))
+        >>> comp = D.Independent(D.Normal(
+                     torch.randn(5,2), torch.rand(5,2)), 1)
+        >>> gmm = MixtureSameFamily(mix, comp)
+
+        # Construct a batch of 3 Gaussian Mixture Models in 2D each
+        # consisting of 5 random weighted bivariate normal distributions
+        >>> mix = D.Categorical(torch.rand(3,5))
+        >>> comp = D.Independent(D.Normal(
+                    torch.randn(3,5,2), torch.rand(3,5,2)), 1)
+        >>> gmm = MixtureSameFamily(mix, comp)
+
+    Args:
+        mixture_distribution: `torch.distributions.Categorical`-like
+            instance. Manages the probability of selecting component.
+            The number of categories must match the rightmost batch
+            dimension of the `component_distribution`. Must have either
+            scalar `batch_shape` or `batch_shape` matching
+            `component_distribution.batch_shape[:-1]`
+        component_distribution: `torch.distributions.Distribution`-like
+            instance. Right-most batch dimension indexes component.
+    """
+    arg_constraints = {}
+    has_rsample = False
+
+    def __init__(self,
+                 mixture_distribution,
+                 component_distribution,
+                 validate_args=None):
+        self._mixture_distribution = mixture_distribution
+        self._component_distribution = component_distribution
+
+        if not isinstance(self._mixture_distribution, Categorical):
+            raise ValueError(" The Mixture distribution needs to be an "
+                             " instance of torch.distribtutions.Categorical")
+
+        if not isinstance(self._component_distribution, Distribution):
+            raise ValueError("The Component distribution need to be an "
+                             "instance of torch.distributions.Distribution")
+
+        # Check that batch size matches
+        mdbs = self._mixture_distribution.batch_shape
+        cdbs = self._component_distribution.batch_shape[:-1]
+        if len(mdbs) != 0 and mdbs != cdbs:
```

Comment:
This should be weaker so we can support broadcasting. One way to write the test would be
```py
for size1, size2 in zip(reversed(mixture_distribution.batch_shape),
                        reversed(component_distribution.batch_shape[:-1])):
    if size1 != 1 and size2 != 1 and size1 != size2:
        raise ...
```
Alternatively you could initially expand the two distributions to have compatible batch shape at the beginning of the `__init__` method; I guess that would be more similar to the `broadcast_all()` handling in other distributions. Something like:
```py
m_shape = mixture_distribution.batch_shape
c_shape = component_distribution.batch_shape
batch_shape = [1] * max(len(m_shape), len(c_shape[:-1]))
for shape in [m_shape, c_shape[:-1]]:
    for i, size in enumerate(reversed(shape)):
        if size != 1:
            batch_shape[-1-i] = size
batch_shape = torch.Size(batch_shape)
if m_shape != batch_shape:
    mixture_distribution = mixture_distribution.expand(batch_shape)
if c_shape[:-1] != batch_shape:
    component_distribution = component_distribution.expand(batch_shape + c_shape[-1:])
...
```